script/modelling/facility_location.jl

```diff
@@ -0,0 +1,262 @@
+for i in 1:m
+    for j in 1:n
+        if x_[i, j] == 1
+           plot!([Xc[i], Xf[j]], [Yc[i], Yf[j]], color=:black, label=nothing)
+        end
+    end
+end
+
+display(p)
+
+
+#' ## Capacitated Facility location
+
+#' ### Problem formulation
+#' 
+#' The capacitated variant introduces a capacity constraint on each facility, i.e., clients have a certain level of demand to be served, while each facility only has finite capacity which cannot be exceeded.
+#' 
+#' Specifically, let
+#' * $a_{i} \geq 0$ denote the demand of client $i$
+#' * $q_{j} \geq 0$ denote the capacity of facility $j$
+#' 
+#' The capacity constraints then write
+#' \begin{align}
+#' \sum_{i} a_{i} x_{i, j} &\leq q_{j} y_{j} && \forall j \in N
+#' \end{align}
+#' 
+#' Note that, if $y_{j}$ is set to $0$, the capacity constraint above automatically forces $x_{i, j}$ to $0$.
+
+#' Thus, the capacitated facility location can be formulated as follows
+#'
+#' \begin{align}
+#' \min_{x, y} \ \ \ &
+#' \sum_{i, j} c_{i, j} x_{i, j} + 
+#' \sum_{j} f_{j} y_{j}\\
+#' s.t. &
+#' \sum_{j} x_{i, j} = 1, && \forall i \in M\\
+#' & \sum_{i} a_{i} x_{i, j} \leq q_{j} y_{j}, && \forall j \in N\\
+#' & x_{i, j}, y_{j} \in \{0, 1\}, && \forall i \in M, j \in N
+#' \end{align}
+#'
+#' For simplicity, we will assume that there is enough capacity to serve the demand,
+#'  i.e., there exists at least one feasible solution.
+
+#+
+# Demands
+a = rand(1:3, m);
+
+# Capacities
+q = rand(5:10, n);
+
+#+
+# Display the data
+scatter(Xc, Yc, label=nothing,
+    markershape=:circle, markercolor=:blue, markersize= 2 .*(2 .+ a)
+)
+
+scatter!(Xf, Yf, label=nothing, 
+    markershape=:rect, markercolor=:white, markersize= q,
+    markerstrokecolor=:red, markerstrokewidth=2
+)
+
+#' ### JuMP implementation
+
+#+
+# Create a JuMP model
+cfl = Model(GLPK.Optimizer)
+
+# Variables
+@variable(cfl, y[1:n], Bin);
+@variable(cfl, x[1:m, 1:n], Bin);
+
+# Each client is served exactly once
+@constraint(cfl, client_service[i in 1:m], sum(x[i, :]) == 1)
+
+# Capacity constraint
+@constraint(cfl, capacity, x'a .<= (q .* y))
+
+# Objective
+@objective(cfl, Min, f'y + sum(c .* x));
+
+#+
+# Solve the problem
+optimize!(cfl)
+println("Optimal value: ", objective_value(cfl))
+
+#' ### Visualizing the solution
+
+#+
+x_ = value.(x) .> 1 - 1e-5
+y_ = value.(y) .> 1 - 1e-5
+
+# Display the solution
+p = scatter(Xc, Yc, label=nothing,
+    markershape=:circle, markercolor=:blue, markersize= 2 .*(2 .+ a)
+)
+
+mc = [(y_[j] ? :red : :white) for j in 1:n]
+scatter!(Xf, Yf, label=nothing, 
+    markershape=:rect, markercolor=mc, markersize=q,
+    markerstrokecolor=:red, markerstrokewidth=2
+)
+
+# Show client-facility assignment
+for i in 1:m
+    for j in 1:n
+        if x_[i, j] == 1
+            plot!([Xc[i], Xf[j]], [Yc[i], Yf[j]], color=:black, label=nothing)
+            break
+        end
+    end
+end
+
+display(p)
+
+
+#' ## Further
+#' * Benders decomposition is a method of choice for solving
```

Comment:
Maybe add a link to the Benders tutorial ? 
https://github.com/JuliaOpt/JuMPTutorials.jl/blob/master/script/optimization_concepts/benders_decomposition.jl